threedangles/src/Vec4.cpp

```diff
@@ -1,72 +1,158 @@
 #include <Vec4.hpp>
 #include <cmath>
+#include <cassert>
+#include <cpu/Vec4CPU.hpp>
+#include <simd/Vec4SSE.hpp>
 
-Vec4::Vec4(const float x, const float y, const float z) : Vec4(x, y, z, 1.0f)
+Vec4::eImpl Vec4::impl = Vec4::eImpl::CPU;
+
+Vec4::Vec4(const float x, const float y, const float z, const float w) :x(x), y(y), z(z), w(w)
 {
+    // This is an overhead for every single Vector...
+    // Must be done at compile time with MACROs so no need of using function pointers
+    // the macro can just put the right namespace (cpu::, sse::) for eg.
+    // cons only at compile time a feature is enabled, can't fallback on CPU without SSE.
+    // unless there is a specifc pattern to do it, 
+    // but anyway it is compiled already with SSE even the CPU implementation due
+    // to compiler optimization
+    // probably can just be removed then.
+    switch (impl)
```

Comment:
resolve it with macro.

keep the same `Vec4` interface, problem is with Cuda that can't use std::vector .... so might need to be done separately like a cudaEngine. ... 

so set up this to use "CPU" by default and comment out this code and remove the function pointer too and just put directly the "old version" of computing it. ... 

or use the inline __fastcall / __vectorcall to the Vec4CPU  ?